express-notes-api/index.js

```diff
@@ -0,0 +1,163 @@
+const fs = require('fs');
+const JSONMiddleware = express.json();
+app.use(JSONMiddleware);
+
+// require automatically parses
+
+app.get('/api/notes', function (req, res) {
+  const array = [];
+  for (const property in data.notes) {
+    array.push(data.notes[property]);
+  }
+  res.status(200);
+  res.send(array);
+});
+
+app.get('/api/notes/:id', function (req, res) {
+  const reqId = Number(req.params.id);
+  if (isNaN(reqId)) {
+    const errorObject = {
+      error: 'id must be a positive integer'
+    };
+    res.status(400);
+    res.send(errorObject);
+    return;
+  }
+  const array = [];
+
+  if (!data.notes[reqId]) {
+    const cannotFindObject = {
+      error: `cannot find note with id ${reqId}`
+    };
+    res.send(cannotFindObject);
+    res.sendStatus(404);
+    return;
+  } else {
+    array.push(data.notes[reqId]);
+  }
+
+  res.send(array[0]);
+
+});
+
+let nextId = data.nextId;
+
+app.post('/api/notes', function (req, res) {
+  const reqBody = req.body;
+  if (!reqBody.content) {
+    const errorObject = {
+      error: 'content is a required field'
+    };
+    res.status(400);
+    res.send(errorObject);
+  } else if (reqBody.content) {
+    reqBody.id = nextId;
+    data.notes[nextId] = reqBody;
+    fs.writeFile('data.json', JSON.stringify(data, null, 2), 'utf-8', function (err) {
+      if (err) {
+        throw err;
+      }
+    });
+    nextId++;
+    res.status(201);
+    res.send(reqBody);
+  }
+});
+
+app.delete('/api/notes/:id', (req, res) => {
+  const reqId = Number(req.params.id);
+
+  if (isNaN(reqId)) {
+    const fourHundrerErrorObject = {
+      error: 'Id must be a positive integer'
+    };
+    res.status(400);
+    res.send(fourHundrerErrorObject);
+  }
+
+  if (!data.notes[reqId]) {
+    const cannotFindObject = {
+      error: `cannot find note with id ${reqId}`
+    };
+    res.status(404);
+    res.send(cannotFindObject);
+
+  } else {
+    delete data.notes[reqId];
+    fs.writeFile('data.json', JSON.stringify(data, null, 2), 'utf-8', function (err) {
+      if (err) {
+        throw err;
+      }
+    });
+    res.sendStatus(200);
+  }
+
+});
+
+app.put('/api/notes/:id', (req, res) => {
+  const reqId = Number(req.params.id);
+  const reqBody = req.body;
+
+  if (isNaN(reqId)) {
+    const fourHundrerErrorObject = {
+      error: 'Id must be a positive integer'
+    };
+    res.status(400);
+    res.send(fourHundrerErrorObject);
+  } else if (!reqBody.content) {
+    const errorObject = {
+      error: 'content is a required field'
+    };
+    res.status(400);
+    res.send(errorObject);
+  } else if (!data.notes[reqId]) {
+    const fourHundredFourErrorObject = {
+      error: `cannot find note with id ${reqId}`
+    };
+    res.status(404);
+    res.send(fourHundredFourErrorObject);
+  } else {
+    data.notes[reqId].content = reqBody.content;
+    fs.writeFile('data.json', JSON.stringify(data, null, 2), 'utf-8', function (err) {
+      if (err) {
+        throw err;
+      }
```

Comment:
Instead of throwing errors, you should log them with `console.error()`. As mentioned in the instructions, you shouldn't `throw` the error. This is because it will crash your server process.